### PR TITLE
feature(users): user registration

### PR DESCRIPTION
**Pull request by MuhanguziDavid**
### What does this PR do?

It enables users to signup using Authors Heaven front end.

### Description of tasks to be completed?

Currently, a user cannot signup to Authors Heaven using the front end
This PR adds functionality to allow users to signup to Authors Heaven using the front end

### How should this be manually tested?

- Run `npm install` to install all dependencies
- Run `npm start`
- Enter the url `http://localhost:3000/signup` in your browser.
- Enter user details in the registration form and click `submit`

### What are the relevant pivotal tracker stories?

[#160609513](https://www.pivotaltracker.com/story/show/160609513)

### Any background context you want to add?

N/A

### Screenshots
registerform

### CODE REVIEW

File: src/actions/registerUserActions.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import axios from 'axios';
+import { toast } from 'react-toastify';
+import { REGISTER_USER_SUCCESS, REGISTER_USER_ERROR } from './types';
+
+export const fetchUsers = (postData) => dispatch => {
```
Comment (RonKbS) [inline src/actions/registerUserActions.js]: Nice work on implementing this feature, @MuhanguziDavid.
This block of code is unclear to me. Is it necessary to have two arrow functions like this?

**Comment by RonKbS** [comment]
The circleci and travis tests are failing, @MuhanguziDavid.
It may be because you have not included some packages in the package.json file.

File: src/actions/registerUserActions.js (hunk 1 above)
Comment (MuhanguziDavid) [reply src/actions/registerUserActions.js]: @RonKbS , yes, it is necessary. Like having one function within another.

File: src/actions/registerUserActions.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  axios
+    .post('http://127.0.0.1:8000/api/users/', postData)
+    .then((response) => {
+      localStorage.setItem('auth_token', response.data.user.auth_token);
+      dispatch({ type: REGISTER_USER_SUCCESS, payload: true });
+      toast.success('Signup successful', { autoClose: false });
```
Comment (mendozabree) [inline src/actions/registerUserActions.js]: Good work @MuhanguziDavid! I am impressed with your test coverage!
You should consider changing the `autoClose` functionality.
Something of the sort `{ autoClose: 3500, hideProgressBar: true }` the latter part removes the default progress bar. The same toast change applies to the `error` toast on line 33.

**Review by mendozabree — COMMENTED** [review]
@MuhanguziDavid, good work with testing the component and the naming conventions. 
Though you missed one, consider renaming the component test file, for example, a component named `CreateArticle.js` would have its corresponding file named `CreateArticle.test.js`

File: src/actions/registerUserActions.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    })
+    .catch((error) => {
+      let emailKey = 'Email: ';
```
Comment (dondrzzy) [inline src/actions/registerUserActions.js]: Please use a for-in loop to check the errors object. [Link](https://stackoverflow.com/questions/684672/how-do-i-loop-through-or-enumerate-a-javascript-object)

File: src/components/register/RegisterUser.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    };
+
+    this.props.fetchUsers(payload);
+  }
+
+  render() {
+    const { email, password, username } = this.state;
+    return (
+      <div className="registerForm">
+        <h1>Register</h1>
```
Comment (dondrzzy) [inline src/components/register/RegisterUser.js]: Get a few ideas from [here](http://getbootstrap.com/docs/4.1/components/forms/)

File: src/components/register/RegisterUser.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          <button className="btn" type="submit" value="Submit">Submit</button>
+        </form>
+      </div>
+    );
+  }
+}
+
+const mapStateToProps = state => ({
+  registerUserSuccess: state.user.registerUserSuccess,
+  registerUserError: state.user.registerUserError,
```
Comment (dondrzzy) [inline src/components/register/RegisterUser.js:92]: Do you still need this?

File: src/actions/registerUserActions.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import axios from 'axios';
+import { toast } from 'react-toastify';
+import { REGISTER_USER_SUCCESS, REGISTER_USER_ERROR } from './types';
+
+export const fetchUsers = (postData) => dispatch => {
+  axios
```
Comment (dondrzzy) [inline src/actions/registerUserActions.js]: Please use the `axios Instance` here. [Link](https://github.com/axios/axios)

File: src/tests/actions/registerUserAction.test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+describe('registerUserAction', () => {
+  let store;
+  let mockAdapter;
+  const flushAllPromises = () => new Promise (resolve => setImmediate(resolve));
+  const emailKey = 'Email: ';
+  const usernameKey = 'Username: ';
+  const passwordKey = 'Password: ';
+
+  beforeEach(() => {
+    mockAdapter = new MockAdapter(axios);
```
Comment (dondrzzy) [inline src/tests/actions/registerUserAction.test.js]: Use mock instead of mockAdapter

File: src/tests/actions/registerUserAction.test.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    const mockStore = configureMockStore();
+    store = mockStore({});
+  });
+
+  it('should register a user', async () => {
```
Comment (dondrzzy) [inline src/tests/actions/registerUserAction.test.js]: This test should be in the directory `tests/actions/userActions`.

File: src/reducers/registerUserReducer.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { REGISTER_USER_SUCCESS, REGISTER_USER_ERROR } from '../actions/types';
```
Comment (dondrzzy) [inline src/reducers/registerUserReducer.js]: `userReducer` not `registerUserReducer`

File: src/actions/registerUserActions.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import axios from 'axios';
```
Comment (dondrzzy) [inline src/actions/registerUserActions.js]: `userActions` not `registerUserActions`

File: src/tests/actions/registerUserAction.test.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      [
+        { type: REGISTER_USER_SUCCESS, payload: true },
+      ],
+    );
+  });
+
+  it('should not register a user without a valid username', async () => {
+    const response_data = {
+      token: 'my secret token',
+      errors: { email: ['This field may not be blank.'], password: ['This field may not be blank.'], username: null },
```
Comment (dondrzzy) [inline src/tests/actions/registerUserAction.test.js]: Email value should never be an array of a string. Get the 0 index value.

File: src/tests/components/registerUser.test.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  it('should pass user data form submission', () => {
+    wrapper.find('#email').simulate('change', { target: { name: 'email', value: 'user100@gmail.com' } });
+    wrapper.find('#password').simulate('change', { target: { name: 'password', value: 'Abc12345' } });
+    wrapper.find('#username').simulate('change', { target: { name: 'username', value: 'user100' } });
+    wrapper.find('#form').simulate('submit', { preventDefault() {} });
+    expect(fetchUsers.mock.calls[2][0]).toEqual({ user: { email: 'user100@gmail.com', password: 'Abc12345', username: 'user100' } });
+  });
+
+  it('should redirect on successful login', () => {
```
Comment (dondrzzy) [inline src/tests/components/registerUser.test.js]: What is this test doing?

**Review by dondrzzy — COMMENTED** [review]
Please refactor

File: src/actions/registerUserActions.js (hunk 2 above)
Comment (MuhanguziDavid) [reply src/actions/registerUserActions.js]: The autoClose functionality has been changed

File: src/components/register/RegisterUser.js (hunk 5 above)
Comment (MuhanguziDavid) [reply src/components/register/RegisterUser.js:92]: I left that there in the event that errors are shown within the form

**Comment by MuhanguziDavid** [comment]
Thanks for the feedback, I have refactored accordingly.

File: src/tests/components/registerUser.test.js (hunk 12 above)
Comment (MuhanguziDavid) [reply src/tests/components/registerUser.test.js]: I had put it there to test whether there is redirection on successful user signup, the test has now been added.

**Pull request merged**